data/transactions/logic/eval.go

```diff
@@ -2191,11 +2192,71 @@ func TxnFieldToTealValue(txn *transactions.Transaction, groupIndex int, field Tx
 	return sv.toTealValue(), err
 }
 
-func (cx *EvalContext) getTxID(txn *transactions.Transaction, groupIndex int) transactions.Txid {
+// currentTxID is a convenience method to get the Txid for the txn being evaluated
+func (cx *EvalContext) currentTxID() transactions.Txid {
+	if cx.Proto.UnifyInnerTxIDs {
+		// can't just return cx.txn.ID() because I might be an inner txn
+		return cx.getTxID(&cx.txn.Txn, cx.groupIndex, false)
+	}
+
+	// original behavior, for backwards comatability
+	return cx.txn.ID()
+}
+
+// getTxIDNotUnified is a backwards-compatible getTxID used when the consensus param UnifyInnerTxIDs
+// is false. DO NOT call directly, and DO NOT change its behavior
+func (cx *EvalContext) getTxIDNotUnified(txn *transactions.Transaction, groupIndex int) transactions.Txid {
+	if cx.EvalParams.txidCache == nil {
+		cx.EvalParams.txidCache = make(map[int]transactions.Txid, len(cx.TxnGroup))
+	}
+
+	txid, ok := cx.EvalParams.txidCache[groupIndex]
+	if !ok {
+		if cx.caller != nil {
+			innerOffset := len(cx.caller.txn.EvalDelta.InnerTxns)
+			txid = txn.InnerID(cx.caller.txn.ID(), innerOffset+groupIndex)
+		} else {
+			txid = txn.ID()
+		}
+		cx.EvalParams.txidCache[groupIndex] = txid
+	}
+
+	return txid
+}
+
+func (cx *EvalContext) getTxID(txn *transactions.Transaction, groupIndex int, inner bool) transactions.Txid {
+	// inner indicates that groupIndex is an index into the most recent inner txn group
+
 	if cx.EvalParams == nil { // Special case, called through TxnFieldToTealValue. No EvalParams, no caching.
 		return txn.ID()
 	}
 
+	if !cx.Proto.UnifyInnerTxIDs {
+		// original behavior, for backwards comatability
+		return cx.getTxIDNotUnified(txn, groupIndex)
+	}
+
+	if inner {
+		// Initialize innerTxidCache if necessary
+		if cx.EvalParams.innerTxidCache == nil {
+			lastGroupLen := len(cx.getLastInnerGroup())
+			cx.EvalParams.innerTxidCache = make(map[int]transactions.Txid, lastGroupLen)
```

Comment:
I'm inclined to not bother using this length. Just take the default, since getInnerGrooup actually has to do some work.